### ISSUE TITLE
CombinedDataset or mixing data of different kinds

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1782]
How to combine multiple different kinds of data (e.g. paired + unpaired)? There is `CombinedDataset` for that, but maybe there are also other better ways?

What’s mostly missing from `CombinedDataset` is to have a fixed ratio in every batch (currently the ratio per batch is arbitrary, only the overall ratio can be specified).

I think that’s actually very simple to add: in `init_seq_order`, we get the `num_seqs` for the current sub epoch for each of the sub datasets. Then in `_collect_single_seq`, we just need to take evenly from each sub dataset. We could do this sort of logic when the user configures `seq_ordering=="evenly"` or so.

The only problem with this approach: there will be a lot of padding. You can use laplace on the sub datasets, but then when seqs are combined in a batch, they still might have very different seq lens.

What other alternatives are there?

There was the idea about having separate batches: this would avoid the padding problem. And then you could use gradient accumulation if necessary.

Or maybe also multi GPU training, and one GPU is on one dataset, the other on the other dataset.

How to implement this?

One idea was: In `_create_data_loader`, we could also check for another config option `custom_create_data_loader`, which is called when set. It gets `orig_create_data_loader` or so as argument. Then, we could somehow wrap multiple train datasets into one, e.g. using `CombinedDataset` or whatever. But our `custom_create_data_loader`, we would check for `CombinedDataset`, and then actually create two data loaders for each of the subsets, and another wrapper data loader around them. That would solve the problem.

Another thing: now with this, you could easily do one batch from one kind of data, next batch the other kind of data. But then at the end, you are likely not finishing both at the same time. You could instead during the iteration look at `complete_frac` or `seq_idx/num_seqs` (you get that info from the batch), and always pick the dataloader which stays behind. That way should make it even again, but not necessarily exactly one batch after the other flipping.

(Small caveat: `complete_frac` or `seq_idx/num_seqs` can be inaccurate e.g. with `online_shuffle_batches` or with `PostProcessingDataset`, or maybe other things.)

Another idea: in the batching (`get_batching_iterable_dataset_from_config`), you could try to make sure that you covered the same amount data in both datasets (again via `complete_frac` or `seq_idx/num_seqs`), and whenever you hit the batch size limits in one of it, you stop for both. Then again you make sure that you have exactly the same amount of batches for both.

Yet another idea: we can still use `CombinedDataset` with `seq_ordering=="evenly"` (as discussed before), and then have custom batching logic, which is already possible right now via setting `torch_batching`. In that custom batching logic, the user can basically implement exactly those variants that i proposed before. I think this is by far the easiest solution. No real changes needed on RETURNN side (except maybe to extend `CombinedDataset` with `seq_ordering=="evenly"`, but that’s simple).

(cc @robin-p-schmitt @Stefanwuu)
